atlassian/bitbucket/cloud/repositories/pullRequests.py

```diff
@@ -0,0 +1,197 @@
+        """ True if the pull request was merged """
+        return self.get_data("state").upper() == "MERGED"
+
+    @property
+    def is_open(self):
+        """ True if the pull request is open """
+        return self.get_data("state").upper() == "OPEN"
+
+    @property
+    def is_superseded(self):
+        """ True if the pull request was superseded """
+        return self.get_data("state").upper() == "SUPERSEDED"
+
+    @property
+    def created_on(self):
+        """ time of creation """
+        return datetime.strptime(self.get_data("created_on"), "%Y-%m-%dT%H:%M:%S.%f%z")
+
+    @property
+    def updated_on(self):
+        """ time of last update """
+        uo_str = self.get_data("updated_on")
+        uo_dt = datetime.strptime(uo_str, "%Y-%m-%dT%H:%M:%S.%f%z") if uo_str else uo_str
+        return uo_dt
+
+    @property
+    def close_source_branch(self):
+        """ close source branch flag """
+        return self.get_data("close_source_branch")
+
+    @property
+    def source_branch(self):
+        """ source branch """
+        return self.get_data("source")["name"]["branch"]
+
+    @property
+    def destination_branch(self):
+        """ destination branch """
+        return self.get_data("destination")["branch"]["name"]
+
+    @property
+    def comment_count(self):
+        """ number of comments """
+        return self.get_data("comment_count")
+
+    @property
+    def task_count(self):
+        """ number of tasks """
+        return self.get_data("task_count")
+
+    @property
+    def declined_reason(self):
+        """ reason for declining """
+        return self.get_data("reason")
+
+    @property
+    def author(self):
+        """ User object of the author """
+        return User(None, self.get_data("author"))
+
+    def participants(self):
+        """ Returns a generator object of participants """
+        for participant in self.get_data("participants"):
+            yield Participant(participant)
+
+        return
+
+    def reviewers(self):
+        """ Returns a generator object of reviewers """
+        for reviewer in self.get_data("reviewers"):
+            yield User(None, reviewer)
+
+        return
+
+    def comment(self, raw_message):
+        """ Commenting the pull request in raw format """
+        markupstrings = ["markdown", "creole", "plaintext"]
+        if not raw_message:
+            raise ValueError("No message set")
+
+        data = {
+            "content": {
+                "raw": raw_message,
+            }
+        }
+
+        return self.post("comments", data)
+
+    def approve(self):
+        """ Approve a pull request if open """
+        self._check_if_open()
+        data = {"approved": True}
+        return self.post("approve", data)
+
+    def unapprove(self):
+        """ Unapporve a pull request if open """
+        self._check_if_open()
+        return self.delete("approve")
+
+    def merge(self, merge_strategy="merge_commit", close_source_branch=None):
```

Comment:
In C we say 'No magic constants'. I prever class static variables for the string. E.g.
```python
class PullRequest(BitbucketCloudBase):
    MERGE_COMMIT = "merge_commit"
    SQUASH = "squash"
    FAST_FORWARD = "fast_forward"

    def merge(self, merge_strategy=PullRequest.MERGE_COMMIT, close_source_branch=None):
```
The string is defined in one place and you get a script error for a typo.

Reply:
Maybe we should generate a `const.py` to place all constants there..

Reply:
But this can result in a big list and you need to comment where the constant is used.
For the permissions in the server implementation I used the local approach.

Reply:
Sure, but thats the reason why we should do this. There is no need to define constants twice. Thats a pretty common way in other projects. We could even add prefix to the constants like PR_ for pull requests or PIPE_ for pipelines etc.

Nevertheless in such a case server should be separated from cloud.

Reply:
`def merge(self, merge_strategy=PullRequest.MERGE_COMMIT, close_source_branch=None):` this would not work. You can't specify constants defined in a class as defaults. Therefore it must be defined outside of the class e.g. on top of the package or as mentioned in a `const.py`. 

That's another reason why we should use it.

Reply:
But from the OO view the constants are part of the module and should not be defined in a global file. If I search the constants for a function I'm searching in the namespace/package of the function and not for a prefixed constant in a global package.
I prefer a package variable and a class variable referencing the package variable to reference the value with the dot notation of the Pull Request instance.